ios/browser/api/history/brave_history_api.mm

```diff
@@ -0,0 +1,309 @@
+    browsing_history_service_ =
+        std::make_unique<history::BrowsingHistoryService>(
+            browsing_history_driver_.get(), history_service_, sync_service);
+
+    web_history_service_ =
+        ios::WebHistoryServiceFactory::GetForBrowserState(browser_state_);
+    DCHECK(history_service_);
+    DCHECK(browsing_history_service_);
+  }
+  return self;
+}
+
+- (void)dealloc {
+  browser_state_ = nil;
+  history_service_ = nil;
+  web_history_service_ = nil;
+}
+
+- (id<HistoryServiceListener>)addObserver:(id<HistoryServiceObserver>)observer {
+  return [[HistoryServiceListenerImpl alloc] init:observer
+                                   historyService:history_service_];
+}
+
+- (void)removeObserver:(id<HistoryServiceListener>)observer {
+  [observer destroy];
+}
+
+- (bool)isLoaded {
+  // Triggers backend to load if it hasn't already, and then returns whether
+  // it's finished loading.
+  return history_service_->BackendLoaded();
+}
+
+- (void)addHistory:(IOSHistoryNode*)history {
+  // Default Add History method with Typed Transition Type
+  // Add History is only called when a tyoped_url is added on iOS side
+  [self addHistory:history pageTransition:BraveHistoryTransitionType_TYPED];
+}
+
+- (void)addHistory:(IOSHistoryNode*)history
+    pageTransition:(BraveHistoryTransitionType)pageTransition {
+  DCHECK_CURRENTLY_ON(web::WebThread::UI);
+  DCHECK(history_service_->backend_loaded());
+
+  history::HistoryAddPageArgs args;
+  args.url = net::GURLWithNSURL(history.url);
+  args.time =
+      base::Time::FromDoubleT([history.dateAdded timeIntervalSince1970]);
+  args.redirects = history::RedirectList();
+  // Important! Only Typed Transtion is being synced in core side
+  args.transition = pageTransition == BraveHistoryTransitionType_TYPED
+                        ? ui::PAGE_TRANSITION_TYPED
+                        : ui::PAGE_TRANSITION_LINK;
+  args.hidden = false;
+  args.visit_source = history::VisitSource::SOURCE_BROWSED;
+  args.consider_for_ntp_most_visited = true;
+  args.title = base::SysNSStringToUTF16(history.title);
+  args.floc_allowed = false;  // Disable Floc - Not allow tracking
+
+  history_service_->AddPage(args);
+}
+
+- (void)removeHistory:(IOSHistoryNode*)history {
+  DCHECK_CURRENTLY_ON(web::WebThread::UI);
+  DCHECK(history_service_->backend_loaded());
+
+  // Deletes a specific URL using history service and web history service
+  history_service_->DeleteLocalAndRemoteUrl(web_history_service_,
+                                            net::GURLWithNSURL(history.url));
+}
+
+- (void)removeAllWithCompletion:(void (^)())completion {
+  DCHECK_CURRENTLY_ON(web::WebThread::UI);
+  DCHECK(history_service_->backend_loaded());
+
+  // Deletes entire History and from all synced devices
+  history_service_->DeleteLocalAndRemoteHistoryBetween(
+      web_history_service_, base::Time::Min(), base::Time::Max(),
+      base::BindOnce(completion), &tracker_);
```

Comment:
This is unsafe because HistoryService necessarily has to outlive BraveHistoryAPI or you're going to have crashes on shutdown. It's safe in history_api.cc because `ExtensionFunction` is ref counted, but you need to use a method with WeakPtr here

Reply:
also this completion callback could easily cause a crash if the caller doesn't use weak references. Rather than do a direct `BindOnce` I would bind it to a method on this class with WeakPtr and pass the completion callback through

Reply:
`weakSelf` which was referenced outside the block is referenced inside and a check is added.